pytact/data_reader.py

```diff
@@ -781,6 +771,16 @@ def is_file_representative(self) -> bool:
         Se also `Dataset.representative`."""
         return self._lreader[self._graph].representative == self.node.nodeid
 
+    @classmethod
+    def _build_clusters(cls, definitions : Iterable[Definition]):
+        seen : set[Node] = set()
+        for definition in definitions:
+            node = definition.node
+            if node in seen: continue
+            cluster = list(definition.cluster)
+            yield cluster
+            seen.update(cluster)
```

Comment:
I don't think you really need this `seen` set. Couldn't you do something like this?
```suggestion
        for repr in definitions:
            yield list((repr, *itertools.takewhile(lambda d: d.cluster_representative == repr)))
```

Reply:
Well, I guess you cannot use this when the iterable is coming from `Dataset.definitions(spine_only=False)`. But maybe you should have a special case for that in `Dataset`? It feels like this logic should not be part of `Definition`.

Reply:
I don't see a problem when the iterator is a sequence

Reply:
I was not sure if I can count on the clusters being grouped together, fine, I will use something like that (after I understand your magical code :-) )

Reply:
itertools.groupby could be useful

Reply:
Well, the problem is mainly that I'm not sure if the list of definitions is encoded in the proper ordering such that
1. Clusters occur next to each other.
2. The representative of a cluster occurs first.
But I guess I can check if that holds.

Also, when you iterate over a Sequence, do you actually consume it? Not sure.

Reply:
When using `itertools.group_by`, (2) is not necessary. About (1), I don't know, it is a question about the dataset.
Ad iterating over a sequence, yes, it should work the same way as iterating over a list -- it gets automatically converted into an iterator.

Reply:
I think that the invariants I mentioned above hold even for `Dataset.defninitions`. But I'll check more after I finish my meeting with Jason.